src/Disks/S3/DiskS3.cpp

```diff
@@ -647,14 +661,11 @@ void DiskS3::remove(const String & path)
         if (metadata.ref_count == 0)
         {
             file.remove();
+
+            keys.reserve(keys.size() + metadata.s3_objects.size());
+
             for (const auto & [s3_object_path, _] : metadata.s3_objects)
-            {
-                /// TODO: Make operation idempotent. Do not throw exception if key is already deleted.
-                Aws::S3::Model::DeleteObjectRequest request;
-                request.SetBucket(bucket);
-                request.SetKey(s3_root_path + s3_object_path);
-                throwIfError(client->DeleteObject(request));
-            }
+                keys.addKey(s3_root_path + s3_object_path);
```

Comment:
Afair for in the single request it can't remove more that 1000 objects. Does aws sdk respect that somehow? Shouldn't we split that into chunks?

Reply:
Good point.

Reply:
Fixed. Split inside in addKey method.